### PR TITLE
Set of fixes for boojum integration

### PR DESCRIPTION
**Pull request by StanislavBreadless**
# What ❔

- Fix max system contract address
- Ensure that only deployments are allowed for L1->L2 transactions
- Fix for L1Messenger
- Fix the EIP1559 checks
- Fix charging for memory overhead
- Fix EOA for delegatecalls
- Add forceDeploy for keccak256 for future use
- Ensure 
- Lots of other comments/minor changes

## Checklist

- [ ] PR title corresponds to the body of PR (we generate changelog entries from PRs).
- [ ] Tests for the changes have been added / updated.
- [ ] Documentation comments have been added / updated.
- [ ] Code has been formatted via `zk fmt` and `zk lint`.

### CODE REVIEW

File: contracts/Constants.sol [hunk 1]
```diff
@@ -32,7 +32,7 @@ address constant ECMUL_SYSTEM_CONTRACT = address(0x07);
 /// 0x01 - ecrecover
 /// 0x02 - sha256
 /// Important! So the constant should be updated if more precompiles are deployed.
-uint256 constant CURRENT_MAX_PRECOMPILE_ADDRESS = uint256(uint160(SHA256_SYSTEM_CONTRACT));
+uint256 constant CURRENT_MAX_PRECOMPILE_ADDRESS = uint256(uint160(ECMUL_SYSTEM_CONTRACT));
```
Comment (vladbochok) [inline contracts/Constants.sol:35]: Let's add a test for this, I feel we can forget about this change once again.

File: contracts/Constants.sol [hunk 2]
```diff
@@ -32,7 +32,7 @@ address constant ECMUL_SYSTEM_CONTRACT = address(0x07);
 /// 0x01 - ecrecover
 /// 0x02 - sha256
```
Comment (vladbochok) [inline contracts/Constants.sol:33]: Comment here should be updated.

File: bootloader/bootloader.yul [hunk 3]
```diff
@@ -3255,7 +3289,7 @@ object "Bootloader" {
                 }
             }
 
-            /// @dev Returns the addition of two unsigned integers, reverting on overflow.
+            /// @dev Returns the subtraction of two unsigned integers, reverting on overflow.
```
Comment (vladbochok) [inline bootloader/bootloader.yul:3292]:
```suggestion
            /// @dev Returns the subtraction of two unsigned integers, reverting on underflow.
```

File: contracts/libraries/SystemContractHelper.sol [hunk 4]
```diff
@@ -270,6 +270,8 @@ library SystemContractHelper {
     function getZkSyncMeta() internal view returns (ZkSyncMeta memory meta) {
         uint256 metaPacked = getZkSyncMetaBytes();
         meta.gasPerPubdataByte = getGasPerPubdataByteFromMeta(metaPacked);
+        meta.heapSize = getHeapSizeFromMeta(metaPacked);
```
Comment (vladbochok) [inline contracts/libraries/SystemContractHelper.sol:273]: I think that the constants for getting heap size and aux heap size are incorrect.

- https://github.com/matter-labs/era-system-contracts/blob/sb-boojum-integration-patch/contracts/libraries/SystemContractsCaller.sol#L42

- `META_HEAP_SIZE_OFFSET = 64`
- `META_AUX_HEAP_SIZE_OFFSET = 96`

While in `era-zkevm_opcode_defs`:
- https://github.com/matter-labs/era-zkevm_opcode_defs/blob/v1.4.1/src/definitions/abi/meta.rs#L17

`aux_heap_size` has offset 64 and `heap_size` has offset 96.

Comment (StanislavBreadless) [reply contracts/libraries/SystemContractHelper.sol:273]: @vladbochok are you sure?

```
result.0[1] = (self.heap_size as u64) | ((self.aux_heap_size as u64) << 32);
```

i.e.

```
result.0[1] = self.heap_size + self.aux_heap_size as u64 *2^32
```

the heap size's offset is 64 and aux heap size is 64 + 32 = 96

**Review by koloz193 — APPROVED** [review]

File: contracts/ContractDeployer.sol [hunk 5]
```diff
@@ -44,7 +44,10 @@ contract ContractDeployer is IContractDeployer, ISystemContract {
         }
 
         // It is an EOA, it is still an account.
-        if (ACCOUNT_CODE_STORAGE_SYSTEM_CONTRACT.getRawCodeHash(_address) == 0) {
+        if (
```
Comment (vladbochok) [inline contracts/ContractDeployer.sol:47]: We should test this functionality too.

File: contracts/DefaultAccount.sol [hunk 6]
```diff
@@ -219,7 +217,7 @@ contract DefaultAccount is IAccount {
         _transaction.processPaymasterInput();
     }
 
-    fallback() external payable {
+    fallback() external payable ignoreInDelegateCall {
```
Comment (vladbochok) [inline contracts/DefaultAccount.sol:220]: Should we write test for this?

**Review by nateawelch — APPROVED** [review]

**Review by vladbochok — APPROVED** [review]

**Review by benceharomi — APPROVED** [review]

**Pull request merged**